KAFKA-9753: A few more metrics to add

guozhangwang opened the pull request:
Instance-level:

* number of alive stream threads

Thread-level: 

* avg / max number of records polled from the consumer per runOnce, INFO
* avg / max number of records processed by the task manager (i.e. across all tasks) per runOnce, INFO

Task-level: 

* number of current buffered records at the moment (i.e. it is just a dynamic gauge), DEBUG.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

guozhangwang committed:
first pass

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9756-loop-tasks-order

guozhangwang committed:
fix unit tests

guozhangwang committed:
add unit tests

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9753-active-process-ratio

guozhangwang committed:
first pass

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9756-loop-tasks-order

guozhangwang committed:
should be INFO

guozhangwang committed:
fix unit tests

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9756-loop-tasks-order

guozhangwang committed:
github comments

guozhangwang committed:
rebase

guozhangwang committed:
unit tests

guozhangwang committed:
add a few more metrics

guozhangwang committed:
rebase from trunk

guozhangwang committed:
Merge branch 'K9753-active-process-ratio' of https://github.com/guozhangwang/kafka into K9753-more-metrics

guozhangwang committed:
rebase from trunk

guozhangwang committed:
rebase from trunk

guozhangwang commented:
@vvcephei @cadonna please let me know wdyt about the proposed ones.

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9753-more-metrics

guozhangwang committed:
add num.threads metric

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -762,6 +762,9 @@ private KafkaStreams(final InternalTopologyBuilder internalTopologyBuilder,
         }
 
+        ClientMetrics.addNumAliveStreamThreadMetric(streamsMetrics, (metricsConfig, now) ->
+            Math.toIntExact(Arrays.stream(threads).filter(thread -> thread.state().isAlive()).count()));
```
Bummer that `Stream.count` forces us to round-trip through a long, but whatever.

cadonna replied:
req: Please add verifications or unit tests to verify the values measured by this metric in `KafkaStreamsTest`.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/internals/metrics/ClientMetrics.java:
```diff
@@ -58,6 +60,8 @@ private ClientMetrics() {}
     private static final String TOPOLOGY_DESCRIPTION_DESCRIPTION =
         "The description of the topology executed in the Kafka Streams client";
     private static final String STATE_DESCRIPTION = "The state of the Kafka Streams client";
+    private static final String ALIVE_STREAM_THREADS_DESCRIPTION = "The number of alive stream threads";
```
Since StreamThreads have a number of states that people are familiar with watching, it might be worth being specific about what "alive" means. In particular, it seems like many will wonder if this just means "running", which it does not.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetrics.java:
```diff
@@ -64,11 +65,15 @@ private ThreadMetrics() {}
     private static final String PROCESS_AVG_LATENCY_DESCRIPTION = "The average process latency";
     private static final String PROCESS_MAX_LATENCY_DESCRIPTION = "The maximum process latency";
+    private static final String PROCESS_AVG_RECORDS_DESCRIPTION = "The average number of records processed  within an iteration";
+    private static final String PROCESS_MAX_RECORDS_DESCRIPTION = "The maximum number of records polled  within an iteration";
```
```suggestion
    private static final String PROCESS_MAX_RECORDS_DESCRIPTION = "The maximum number of records processed within an iteration";
```

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetrics.java:
```diff
@@ -64,11 +65,15 @@ private ThreadMetrics() {}
     private static final String PROCESS_RATE_DESCRIPTION = RATE_DESCRIPTION + PROCESS_DESCRIPTION;
     private static final String PROCESS_AVG_LATENCY_DESCRIPTION = "The average process latency";
     private static final String PROCESS_MAX_LATENCY_DESCRIPTION = "The maximum process latency";
+    private static final String PROCESS_AVG_RECORDS_DESCRIPTION = "The average number of records processed  within an iteration";
```
```suggestion
    private static final String PROCESS_AVG_RECORDS_DESCRIPTION = "The average number of records processed within an iteration";
```

vvcephei reviewed (commented):
Thanks for the PR, @guozhangwang ! 

Looks good overall; I just had a few comments.

Can you add some tests verifying the new metrics are registered and work properly?

Thanks,
John

cadonna commented on streams/src/main/java/org/apache/kafka/streams/internals/metrics/ClientMetrics.java:
```diff
@@ -113,4 +117,25 @@ public static void addStateMetric(final StreamsMetricsImpl streamsMetrics,
         );
     }
+
+    public static void addNumAliveStreamThreadMetric(final StreamsMetricsImpl streamsMetrics,
+                                                     final Gauge<Integer> stateProvider) {
+        streamsMetrics.addClientLevelMutableMetric(
+            ALIVE_STREAM_THREADS,
+            ALIVE_STREAM_THREADS_DESCRIPTION,
+            RecordingLevel.INFO,
+            stateProvider
+        );
+    }
+
+    public static void addNumAliveCleanupThreadMetric(final StreamsMetricsImpl streamsMetrics,
+                                                      final Gauge<Integer> stateProvider) {
+        streamsMetrics.addClientLevelMutableMetric(
+            ALIVE_CLEANUP_THREADS,
+            ALIVE_CLEANUP_THREADS_DESCRIPTION,
+            RecordingLevel.INFO,
+            stateProvider
+        );
+    }
```
req: Please add unit tests to `ClientMetricsTest` for this methods.

guozhangwang replied:
Ack.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/TaskMetrics.java:
```diff
@@ -117,6 +120,21 @@ public static Sensor activeProcessRatioSensor(final String threadId,
+        );
+        return sensor;
+    }
+
```
req: Please add unit tests to `TaskMetricsTest` for this methods.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/TaskMetrics.java:
```diff
@@ -117,6 +120,21 @@ public static Sensor activeProcessRatioSensor(final String threadId,
+    public static Sensor activeBufferedRecordsSensor(final String threadId,
+                                                     final String taskId,
+                                                     final StreamsMetricsImpl streamsMetrics) {
+        final String name = ACTIVE_TASK_PREFIX + BUFFERED + StreamsMetricsImpl.RECORDS_SUFFIX;
```
prop: For the suppression buffer, we used `-count` as a suffix instead of `-records` to measure the number of records.

guozhangwang replied:
Ack, that makes sense!

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -628,6 +630,7 @@ public void recordProcessBatchTime(final long processBatchTime) {
 
     @Override
     public void recordProcessTimeRatio(final long allTaskProcessMs) {
```
prop: Could you please rename this method to reflect the new recording inside?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/TaskMetrics.java:
```diff
@@ -82,8 +82,11 @@ private TaskMetrics() {}
+
     private static final String PROCESS_RATIO_DESCRIPTION = "The fraction of time the thread spent " +
         "on processing this task among all assigned active tasks";
+    private static final String NUM_BUFFERED_RECORDS_DESCRIPTION = "Number of buffered records for this active task";
```
req: When reading this description, it is not immediately clear where those records are buffered. Could you please add some reference to the input or similar?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetrics.java:
```diff
@@ -175,6 +180,40 @@ public static Sensor processLatencySensor(final String threadId,
         return sensor;
     }
 
+    public static Sensor pollRecordsSensor(final String threadId,
+                                           final StreamsMetricsImpl streamsMetrics) {
+        final Sensor sensor =
+            streamsMetrics.threadLevelSensor(threadId, POLL + RECORDS_SUFFIX, RecordingLevel.INFO);
+        final Map<String, String> tagMap = streamsMetrics.threadLevelTagMap(threadId);
+        final String threadLevelGroup = threadLevelGroup(streamsMetrics);
+        addAvgAndMaxToSensor(
+            sensor,
+            threadLevelGroup,
+            tagMap,
+            POLL + RECORDS_SUFFIX,
+            POLL_AVG_RECORDS_DESCRIPTION,
+            POLL_MAX_RECORDS_DESCRIPTION
+        );
+        return sensor;
+    }
+
+    public static Sensor processRecordsSensor(final String threadId,
+                                              final StreamsMetricsImpl streamsMetrics) {
+        final Sensor sensor =
+            streamsMetrics.threadLevelSensor(threadId, PROCESS + RECORDS_SUFFIX, RecordingLevel.INFO);
+        final Map<String, String> tagMap = streamsMetrics.threadLevelTagMap(threadId);
+        final String threadLevelGroup = threadLevelGroup(streamsMetrics);
+        addAvgAndMaxToSensor(
+            sensor,
+            threadLevelGroup,
+            tagMap,
+            PROCESS + RECORDS_SUFFIX,
+            PROCESS_AVG_RECORDS_DESCRIPTION,
+            PROCESS_MAX_RECORDS_DESCRIPTION
+        );
+        return sensor;
+    }
+
```
req: Please add unit tests in `ThreadMetricsTest`.

guozhangwang commented:
Okay, if you do not have any objections on the newly proposed sensors I'll update the KIP and send out an email in the thread as well.

cadonna reviewed (commented):
@guozhangwang Thank you for the PR

Here my feedback

guozhangwang committed:
github comments

guozhangwang committed:
rebase from trunk

guozhangwang commented:
@cadonna @vvcephei addressed your comments, please review again.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/internals/metrics/ClientMetricsTest.java:
```diff
@@ -91,6 +91,42 @@ public void shouldAddStateMetric() {
         verify(streamsMetrics);
     }
 
+    @Test
+    public void shouldAddAliveStreamThreadsMetric() {
+        final String name = "alive-stream-threads";
+        final String description = "The current number of alive stream threads that are running or participating in rebalance";
+        final Gauge<Integer> stateProvider = (config, now) -> 1;
+        streamsMetrics.addClientLevelMutableMetric(
+            eq(name),
+            eq(description),
+            eq(RecordingLevel.INFO),
+            eq(stateProvider)
+        );
+        replay(streamsMetrics);
+
+        ClientMetrics.addNumAliveStreamThreadMetric(streamsMetrics, stateProvider);
+
+        verify(streamsMetrics);
+    }
+
+    @Test
+    public void shouldAddAliveCleanupThreadsMetric() {
+        final String name = "alive-cleanup-threads";
+        final String description = "The current number of alive local store directory cleanup threads";
+        final Gauge<Integer> stateProvider = (config, now) -> 1;
+        streamsMetrics.addClientLevelMutableMetric(
+            eq(name),
+            eq(description),
+            eq(RecordingLevel.INFO),
+            eq(stateProvider)
+        );
+        replay(streamsMetrics);
+
+        ClientMetrics.addNumAliveCleanupThreadMetric(streamsMetrics, stateProvider);
+
+        verify(streamsMetrics);
+    }
+
```
prop: Could you please add a private method like `setUpAndVerifyMetric()` for mutable metrics to avoid code duplications? Best would be to rename `setUpAndVerifyMetric()` to `setUpAndVerifyImmutableMetric()` and to call the new method `setUpAndVerifyMutableMetric()`.

guozhangwang replied:
Good call! Will do.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -192,6 +192,10 @@ private void prepareStreams() throws Exception {
         ClientMetrics.addTopologyDescriptionMetric(anyObject(StreamsMetricsImpl.class), anyString());
         ClientMetrics.addStateMetric(anyObject(StreamsMetricsImpl.class), anyObject());
+        ClientMetrics.addNumAliveStreamThreadMetric(anyObject(StreamsMetricsImpl.class), anyObject());
+        EasyMock.expectLastCall().anyTimes();
```
req: We expect the addition of the metric exactly once. Could you please remove line 196?

guozhangwang replied:
Actually this is added intentionally since there are three tests in which we only created the instance without calling its `start()` function -- thinking about it, I will move the registration logic back to constructor and then we can remove it.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -192,6 +192,10 @@ private void prepareStreams() throws Exception {
+        ClientMetrics.addNumAliveStreamThreadMetric(anyObject(StreamsMetricsImpl.class), anyObject());
+        EasyMock.expectLastCall().anyTimes();
+        ClientMetrics.addNumAliveCleanupThreadMetric(anyObject(StreamsMetricsImpl.class), anyObject());
+        EasyMock.expectLastCall().anyTimes();
```
req: same as above

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -445,12 +447,14 @@ public StreamThread(final Time time,
         this.streamsMetrics = streamsMetrics;
         this.commitSensor = ThreadMetrics.commitSensor(threadId, streamsMetrics);
         this.pollSensor = ThreadMetrics.pollSensor(threadId, streamsMetrics);
+        this.pollRecordsSensor = ThreadMetrics.pollRecordsSensor(threadId, streamsMetrics);
```
req: Could you please add verifications of the creation of those metrics in `StreamThreadTest#shouldCreateMetricsAtStartup()`?

guozhangwang replied:
Ack.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -445,12 +447,14 @@ public StreamThread(final Time time,
+        this.pollRecordsSensor = ThreadMetrics.pollRecordsSensor(threadId, streamsMetrics);
+        this.pollRatioSensor = ThreadMetrics.pollRatioSensor(threadId, streamsMetrics);
         this.processLatencySensor = ThreadMetrics.processLatencySensor(threadId, streamsMetrics);
+        this.processRecordsSensor = ThreadMetrics.processRecordsSensor(threadId, streamsMetrics);
```
req: Could you please add verifications of the creation of those metrics in `StreamThreadTest#shouldCreateMetricsAtStartup()`?

guozhangwang replied:
Ack.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -137,6 +138,7 @@ public StreamTask(final TaskId id,
         processRatioSensor = TaskMetrics.activeProcessRatioSensor(threadId, taskId, streamsMetrics);
         processLatencySensor = TaskMetrics.processLatencySensor(threadId, taskId, streamsMetrics);
         punctuateLatencySensor = TaskMetrics.punctuateSensor(threadId, taskId, streamsMetrics);
+        bufferedRecordsSensor = TaskMetrics.activeBufferedRecordsSensor(threadId, taskId, streamsMetrics);
```
req: Please verify the creation of those metrics in `StreamTaskTest#shouldConstructMetricsWithBuiltInMetricsVersionLatest()`.

guozhangwang replied:
Ack.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -381,7 +381,7 @@ public void shouldRecordProcessRatio() {
 
```
req: Please rename this unit test to reflect the new verifications.

guozhangwang replied:
I will add a new function to verify the buffer size.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -381,7 +381,7 @@ public void shouldRecordProcessRatio() {
-        task.recordProcessTimeRatio(100L);
+        task.recordProcessTimeRatioAndBufferSize(100L);
 
         assertEquals(0.25d, (double) metric.metricValue(), 0.0001d);
```
prop: I find `assertThat()` better readable than `assertEquals()`.

guozhangwang replied:
Fair enough :)

cadonna commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -773,14 +776,21 @@ private KafkaStreams(final InternalTopologyBuilder internalTopologyBuilder,
             final Thread thread = new Thread(r, clientId + "-CleanupThread");
             thread.setDaemon(true);
+
+            ClientMetrics.addNumAliveCleanupThreadMetric(streamsMetrics, (metricsConfig, now) -> thread.isAlive() ? 1 : 0);
```
req: Please add verifications or unit tests to verify the values measured by this metric in `KafkaStreamsTest`.

guozhangwang replied:
The `KafkaStreamsTest` has mocks on all of its modules including metrics, I'll do this verification in `MetricsIntegrationTest`.

cadonna reviewed (commented):
@guozhangwang I made a second pass.

We need also to verify the metrics in `MetricsIntegrationTest`.

Here my other comments.

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9753-more-metrics

guozhangwang committed:
bruno's comments

guozhangwang commented:
@cadonna @vvcephei  while addressing your comments I realized that for cleanup, the scheduler may create more than one thread (it would potentially create a new thread whenever it triggers) which means that metric would be like 1 - 0 - 1 - 0 - 1 and not be very useful for monitoring.. so I've decided to remove it, will update the KIP as well.

The PR is now ready for another look.

guozhangwang committed:
remove unused imports

guozhangwang committed:
resolve conflicts from trunk

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9753-more-metricst

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/MetricsIntegrationTest.java:
```diff
@@ -193,30 +204,6 @@
-    private static final String COMPACTION_TIME_MAX = "compaction-time-max";
-    private static final String NUMBER_OF_OPEN_FILES = "number-open-files";
-    private static final String NUMBER_OF_FILE_ERRORS = "number-file-errors-total";
-
```
Thank you for this! I forgot to remove them.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -434,6 +456,20 @@ private void testMetrics(final String builtInMetricsVersion) {
             builtInMetricsVersion
         ));
 
+        assertNotNull(getMetric(
+            "active-process",
+            "%s-ratio",
+            task.id().toString(),
+            builtInMetricsVersion
+        ));
+
```
req: You verify this metric twice, once here in once in `testMetricsForBuiltInMetricsVersionLatest()`. Since it is a newly added metric und you account for the changes in the tags in `getMetric()` I guess you just need to keep the one here.

guozhangwang replied:
Sounds good!

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -375,24 +375,46 @@ public void shouldProcessInOrder() {
+    }
+
+    @Test
+    public void shouldRecordBufferedRecords() {
```
req: The names of this method and the previous method should be switched.

guozhangwang replied:
Ack.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/MetricsIntegrationTest.java:
```diff
@@ -246,6 +233,7 @@ public void before() throws InterruptedException {
         streamsConfiguration.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.String().getClass());
         streamsConfiguration.put(StreamsConfig.METRICS_RECORDING_LEVEL_CONFIG, Sensor.RecordingLevel.DEBUG.name);
         streamsConfiguration.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 10 * 1024 * 1024L);
+        streamsConfiguration.put(StreamsConfig.NUM_STREAM_THREADS_CONFIG, 2);
```
req: Please use a constant for the number of stream threads here and on line 260.

guozhangwang replied:
Ack.

cadonna reviewed (approved):
@guozhangwang LGTM

Just a two trivial requests

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9753-more-metrics

guozhangwang committed:
address more comments

vvcephei reviewed (approved):
Thanks, @guozhangwang !
